File: multi_process_sp/src/main/java/cn/qingtui/multiprocesssp/lib/SharedPreferencesProvider.java
package cn.qingtui.multiprocesssp.lib;

import android.content.*;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * SP的provider，被远程调用
 * Date:2019-05-30
 *
 * @author KongMing
 */
public class SharedPreferencesProvider extends ContentProvider implements SharedPreferences.OnSharedPreferenceChangeListener {
    static String broadcastAction;
    Map<Integer, String> sharedPreferencesMap = new HashMap<>();

    @Override
    public boolean onCreate() {
        return true;
    }

    private SharedPreferences getSharedPreferences(String spName, int mode) {
        SharedPreferences sharedPreferences = getContext().getSharedPreferences(spName, mode);
        sharedPreferencesMap.put(sharedPreferences.hashCode(), spName);
        sharedPreferences.registerOnSharedPreferenceChangeListener(this);
        return sharedPreferences;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        MatrixCursor cursor = new MatrixCursor(projection, 3);

        String spName = selectionArgs[0];
        int mode = Integer.parseInt(selectionArgs[1]);
        SharedPreferences sharedPreferences = getSharedPreferences(spName, mode);
        if (selectionArgs.length > 2) {
            String key = selectionArgs[2];
            String valueClassName = selectionArgs[3];
            String defaultValue = selectionArgs[4];
            findOne(cursor, sharedPreferences, key, valueClassName, defaultValue);
        } else {
            Map<String, ?> map = sharedPreferences.getAll();
            if (map != null && map.entrySet() != null) {
                for (Map.Entry<String, ?> stringEntry : map.entrySet()) {
                    Object value = null;
                    if (stringEntry.getValue() instanceof Set) {
                        value = StringUtils.fromSet((Set<String>) stringEntry.getValue());
                    } else {
                        value = stringEntry.getValue();
                    }
                    cursor.newRow().add(stringEntry.getKey())
                            .add(value.getClass().getName())
                            .add(String.valueOf(value));
                }
            }
        }
        return cursor;
    }

    private void findOne(MatrixCursor cursor, SharedPreferences sharedPreferences, String key, String valueClassName, String defaultValue) {
        Object value = null;
        if (Integer.class.getName().equals(valueClassName)) {
            value = sharedPreferences.getInt(key, Integer.valueOf(defaultValue));
        } else if (Long.class.getName().equals(valueClassName)) {
            value = sharedPreferences.getLong(key, Long.valueOf(defaultValue));
        } else if (Float.class.getName().equals(valueClassName)) {
            value = sharedPreferences.getFloat(key, Float.valueOf(defaultValue));
        } else if (String.class.getName().equals(valueClassName)) {
            value = sharedPreferences.getString(key, String.valueOf(defaultValue));
        } else if (Boolean.class.getName().equals(valueClassName)) {
            value = sharedPreferences.getBoolean(key, Boolean.valueOf(defaultValue));
        } else if (Set.class.getName().equals(valueClassName)) {
            Set vSet = sharedPreferences.getStringSet(key, StringUtils.toSet(defaultValue));
            value = StringUtils.fromSet(vSet);
        }
        cursor.newRow().add(key).add(valueClassName).add(value != null ? String.valueOf(value) : null);
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return "xml/multi.process.sp";
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        String spName = values.getAsString("spName");
        int mode = values.getAsInteger("mode");

        String key = values.getAsString("key");
        String valueClassName = values.getAsString("valueClass");
        Object value = values.get("value");

        SharedPreferences sharedPreferences = getSharedPreferences(spName, mode);

        if (Integer.class.getName().equals(valueClassName)) {
            sharedPreferences.edit().putInt(key, (Integer) value);

        } else if (Long.class.getName().equals(valueClassName)) {
            sharedPreferences.edit().putLong(key, (Long) value);

        } else if (Float.class.getName().equals(valueClassName)) {
            sharedPreferences.edit().putFloat(key, (Float) value);

        } else if (String.class.getName().equals(valueClassName)) {
            sharedPreferences.edit().putString(key, (String) value);

        } else if (Boolean.class.getName().equals(valueClassName)) {
            sharedPreferences.edit().putBoolean(key, (Boolean) value);

        } else if (Set.class.getName().equals(valueClassName)) {
            sharedPreferences.edit().putStringSet(key, StringUtils.toSet((String) value));
        }
        return uri;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        String spName = selectionArgs[0];
        int mode = Integer.parseInt(selectionArgs[1]);
        SharedPreferences sharedPreferences = getSharedPreferences(spName, mode);
        if (selectionArgs.length > 2) {
            String key = selectionArgs[2];
            sharedPreferences.edit().remove(key);
        } else {
            sharedPreferences.edit().clear();
        }
        return 1;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        String spName = selectionArgs[0];
        int mode = Integer.parseInt(selectionArgs[1]);
        Boolean commit = Boolean.valueOf(selectionArgs[2]);

        SharedPreferences sharedPreferences = getSharedPreferences(spName, mode);
        int code;
        if (commit) {
            code = sharedPreferences.edit().commit() ? 1 : 0;
        } else {
            sharedPreferences.edit().apply();
            code = 1;
        }
        return code;
    }


    private void sendChangeBroadcast(String spName, String key) {
        Intent intent = new Intent(getBroadcastAction(getContext()));
        intent.putExtra("spName", spName);
        intent.putExtra("key", key);
        getContext().sendBroadcast(intent);
    }


    public static String getBroadcastAction(Context context) {
        if (broadcastAction != null) {
            return broadcastAction;
        }
        broadcastAction = StringUtils.getSPProviderAuthor(context) + ".broadcast";
        return broadcastAction;
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        String spName = sharedPreferencesMap.get(sharedPreferences.hashCode());
        if (spName != null) {
            sendChangeBroadcast(spName, key);
        }
    }
}

File: multi_process_sp/src/main/java/cn/qingtui/multiprocesssp/lib/SharedPreferencesProvider.kt
package cn.qingtui.multiprocesssp.lib

import android.content.*
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import java.util.*

/**
 * SP的provider，被远程调用
 * Date:2019-05-30
 *
 * @author KongMing
 */
class SharedPreferencesProvider : ContentProvider(),
    OnSharedPreferenceChangeListener {
    var spInfoSet: MutableSet<SPInfo> = HashSet()

    override fun onCreate(): Boolean {
        return true
    }

    private fun getSPInfo(spName: String, mode: Int): SPInfo {
        return spInfoSet.find { it.name == spName }
            ?: SPInfo(spName, context.getSharedPreferences(spName, mode)).also {
                spInfoSet.add(it)
                it.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
            }
    }

    private fun getSharedPreferences(spName: String, mode: Int): SharedPreferences {
        return spInfoSet.find { it.name == spName }?.sharedPreferences
            ?: context.getSharedPreferences(spName, mode).also {
                spInfoSet.add(SPInfo(spName, it))
                it.registerOnSharedPreferenceChangeListener(this)
            }
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ): Cursor? {
        val cursor = MatrixCursor(projection, 3)
        val spName = selectionArgs!![0]
        val mode = selectionArgs[1].toInt()
        val sharedPreferences = getSharedPreferences(spName, mode)
        if (selectionArgs.size > 2) {
            val key = selectionArgs[2]
            val valueClassName = selectionArgs[3]
            val defaultValue = selectionArgs[4]
            findOne(cursor, sharedPreferences, key, valueClassName, defaultValue)
        } else {
            val map = sharedPreferences.all
            if (map?.entries != null) {
                for ((key, value1) in map) {
                    var value: Any? = null
                    value = if (value1 is Set<*>) {
                        StringUtils.fromSet(value1 as Set<String?>?)
                    } else {
                        value1
                    }
                    cursor.newRow().add(key)
                        .add(value!!.javaClass.name)
                        .add(value.toString())
                }
            }
        }
        return cursor
    }

    private fun findOne(
        cursor: MatrixCursor,
        sharedPreferences: SharedPreferences,
        key: String,
        valueClassName: String,
        defaultValue: String
    ) {
        var value: Any? = null
        if (Int::class.java.name == valueClassName) {
            value = sharedPreferences.getInt(key, Integer.valueOf(defaultValue))
        } else if (Long::class.java.name == valueClassName) {
            value = sharedPreferences.getLong(key, java.lang.Long.valueOf(defaultValue))
        } else if (Float::class.java.name == valueClassName) {
            value = sharedPreferences.getFloat(key, java.lang.Float.valueOf(defaultValue))
        } else if (String::class.java.name == valueClassName) {
            value = sharedPreferences.getString(key, defaultValue)
        } else if (Boolean::class.java.name == valueClassName) {
            value = sharedPreferences.getBoolean(key, java.lang.Boolean.valueOf(defaultValue))
        } else if (MutableSet::class.java.name == valueClassName) {
            val vSet: Set<*>? = sharedPreferences.getStringSet(
                key,
                StringUtils.toSet(defaultValue) as? MutableSet<String>?
            )
            value = StringUtils.fromSet(vSet as? MutableSet<String>?)
        }
        cursor.newRow().add(key).add(valueClassName)
            .add(value?.toString())
    }

    override fun getType(uri: Uri): String? {
        return "xml/multi.process.sp"
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val spName = values!!.getAsString("spName")
        val mode = values.getAsInteger("mode")
        val key = values.getAsString("key")
        val valueClassName = values.getAsString("valueClass")
        val value = values["value"]
        val edit = getSPInfo(spName, mode).edit

        if (Int::class.java.name == valueClassName) {
            edit.putInt(key, (value as Int))
        } else if (Long::class.java.name == valueClassName) {
            edit.putLong(key, (value as Long))
        } else if (Float::class.java.name == valueClassName) {
            edit.putFloat(key, (value as Float))
        } else if (String::class.java.name == valueClassName) {
            edit.putString(key, value as? String)
        } else if (Boolean::class.java.name == valueClassName) {
            edit.putBoolean(key, (value as Boolean))
        } else if (MutableSet::class.java.name == valueClassName) {
            edit.putStringSet(
                key,
                StringUtils.toSet(value as? String) as? MutableSet<String>?
            )
        }
        return uri
    }

    override fun delete(
        uri: Uri,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        val spName = selectionArgs!![0]
        val mode = selectionArgs[1].toInt()
        if (selectionArgs.size > 2) {
            val key = selectionArgs[2]
            getSPInfo(spName, mode).edit.remove(key)
        } else {
            getSPInfo(spName, mode).edit.clear()
        }
        return 1
    }

    override fun update(
        uri: Uri,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        val spName = selectionArgs!![0]
        val mode = selectionArgs[1].toInt()
        val commit = java.lang.Boolean.valueOf(selectionArgs[2])
        val sharedPreferences = getSharedPreferences(spName, mode)
        val code: Int
        code = if (commit) {
            if (getSPInfo(spName, mode).edit.commit()) 1 else 0
        } else {
            getSPInfo(spName, mode).edit.apply()
            1
        }
        return code
    }

    private fun sendChangeBroadcast(spName: String, key: String) {
        val intent =
            Intent(getBroadcastAction(context))
        intent.putExtra("spName", spName)
        intent.putExtra("key", key)
        context.sendBroadcast(intent)
    }

    override fun onSharedPreferenceChanged(
        sharedPreferences: SharedPreferences,
        key: String
    ) {
        spInfoSet.find { it.sharedPreferences == sharedPreferences }?.also {
            sendChangeBroadcast(it.name, key)
        }
    }

    companion object {
        var broadcastAction: String? = null

        @JvmStatic
        fun getBroadcastAction(context: Context?): String? {
            if (broadcastAction != null) {
                return broadcastAction
            }
            broadcastAction =
                StringUtils.getSPProviderAuthor(context) + ".broadcast"
            return broadcastAction
        }
    }
}